### PR TITLE
[Seq2Seq Trainer] Make sure padding is implemented for models without pad_token

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR adds padding for models without padding token as well. The logic is the following:

1) If model predicts `targets` < `max_length` => model has to have at least an `eos_token_id`. If model has no `config.pad_token_id` defined than the model simply uses the `config.eos_token_id` for padding.

2) If the model has no `config.eos_token_id`, => model cannot generate predictions shorter than `max_length`. In this case padding will never happen.

@sshleifer @patil-suraj - you guys were right -> the `Trainer` requires padding in any case (also if model has no padding token).
Could you guys review this PR and see if these fixes in Seq2Seq Trainer are ok for you?

### CODE REVIEW

**Comment by patil-suraj** [comment]
LGTM!

File: examples/seq2seq/seq2seq_trainer.py [hunk 1]
```diff
@@ -60,6 +60,11 @@ def __init__(self, config=None, data_args=None, *args, **kwargs):
                 self.config.pad_token_id is not None
             ), "Make sure that `config.pad_token_id` is correcly defined when ignoring `pad_token` for loss calculation or doing label smoothing."
 
+        if self.config.pad_token_id is None and self.config.eos_token_id is not None:
+            logger.warn(
+                f"The `config.pad_token_id` is `None`. Use `config.eos_token_id` = {self.config.eos_token_id} for padding."
```
Comment (patrickvonplaten) [inline examples/seq2seq/seq2seq_trainer.py:65]:
```suggestion
                f"The `config.pad_token_id` is `None`. Using `config.eos_token_id` = {self.config.eos_token_id} for padding.."
```

**Comment by patil-suraj** [comment]
@patrickvonplaten , @sshleifer 
I am seeing a major slowdown on TPU V3-8,  
last time `sshleifer/student_marian_en_ro_6_3` finished 1 epoch in ~6 mins,
now on this branch it's showing ~1hr 20 mins

**Comment by patrickvonplaten** [comment]
> @patrickvonplaten , @sshleifer
> I am seeing a major slowdown on TPU V3-8,
> last time ([9e68d07](https://github.com/huggingface/transformers/commit/9e68d075a4100906509170498480823e7e61874a)) `sshleifer/student_marian_en_ro_6_3` finished 1 epoch in ~6 mins,
> now on this branch it's showing ~1hr 20 mins

Ohh, can you narrow down the commit that caused the slow-down? I took a look again at https://github.com/huggingface/transformers/pull/7809/files and this line I added could be problematic `inputs = copy.deepcopy(inputs)`.

**Comment by patrickvonplaten** [comment]
> > @patrickvonplaten , @sshleifer
> > I am seeing a major slowdown on TPU V3-8,
> > last time ([9e68d07](https://github.com/huggingface/transformers/commit/9e68d075a4100906509170498480823e7e61874a)) `sshleifer/student_marian_en_ro_6_3` finished 1 epoch in ~6 mins,
> > now on this branch it's showing ~1hr 20 mins
> 
> Ohh, can you narrow down the commit that caused the slow-down? I took a look again at https://github.com/huggingface/transformers/pull/7809/files and this line I added could be problematic `inputs = copy.deepcopy(inputs)`.

Yeah this line is actually called at every step -> can you check whether removing the `copy` operation speeds the seq2seq trainer up again? I've been a bit sloppy there I think :-/

**Comment by patil-suraj** [comment]
It's still very slow even after removing that line. I'll try to find the exact commit which casing this slowdown.

File: examples/seq2seq/seq2seq_trainer.py [hunk 2]
```diff
@@ -60,6 +59,11 @@ def __init__(self, config=None, data_args=None, *args, **kwargs):
                 self.config.pad_token_id is not None
             ), "Make sure that `config.pad_token_id` is correcly defined when ignoring `pad_token` for loss calculation or doing label smoothing."
 
+        if self.config.pad_token_id is None and self.config.eos_token_id is not None:
+            logger.warn(
+                f"The `config.pad_token_id` is `None`. Using `config.eos_token_id` = {self.config.eos_token_id} for padding.."
+            )
```
Comment (sshleifer) [inline examples/seq2seq/seq2seq_trainer.py:65]: what if `eos_token_id is None`? Should we raise?

File: examples/seq2seq/seq2seq_trainer.py [hunk 3]
```diff
@@ -178,21 +180,22 @@ def prediction_step(
         """
         inputs = self._prepare_inputs(inputs)
 
+        pred_kwargs = {
```
Comment (sshleifer) [inline examples/seq2seq/seq2seq_trainer.py:183]: why not `gen_kwargs`?

File: examples/seq2seq/seq2seq_trainer.py [hunk 4]
```diff
@@ -205,13 +208,22 @@ def prediction_step(
...
 
     def _pad_tensors_to_max_len(self, tensor, max_length):
-        padded_tensor = self.config.pad_token_id * torch.ones(
+        # If PAD token is not defined at least EOS token has to be defined
+        pad_token_id = self.config.pad_token_id if self.config.pad_token_id is not None else self.config.eos_token_id
+
+        if pad_token_id is None:
+            raise ValueError(
+                f"Make sure that either `config.pad_token_id` or `config.eos_token_id` is defined if tensor has to be padded to `max_length`={max_length}"
+            )
```
Comment (sshleifer) [inline examples/seq2seq/seq2seq_trainer.py:223]: Should we check in `__init__` for faster failures?

**Review by sshleifer — COMMENTED** [review]
LGTM, would prefer stuff moved to init, but don't feel strongly.

File: examples/seq2seq/seq2seq_trainer.py (hunk 2 above)
Comment (sshleifer) [reply examples/seq2seq/seq2seq_trainer.py:65]: should we raise early in that case?

Comment (patrickvonplaten) [reply examples/seq2seq/seq2seq_trainer.py:65]: What I meant is that there are models, like `openai-gpt` or `ctrl` that do not have a `eos_token_id` nor do they have a `pad_token_id` => the way it is implemented now these models could still make use of seq2seqTrainer because they would never require padding (because they never finish early). So I'd just leave it as it is - or if you think that models that don't have an EOS token should not use Seq2SeqTrainer we could raise as well - up to you!

Comment (sshleifer) [reply examples/seq2seq/seq2seq_trainer.py:65]: Didn't understand that they always go to `max_length` your implem makes total sense. Thanks for clarifying.
